feat(terminal)!: cursor shape and blink

gpanders opened the pull request:
When a terminal application running inside the terminal emulator sets the cursor shape or blink status of the cursor, update the cursor in the parent terminal to match.

This removes the "virtual cursor" that has been in use by the terminal emulator since the beginning. The original rationale for using the virtual cursor was to avoid having to support additional UI methods to change the cursor color for other (non-TUI) UIs, instead relying on the TermCursor and TermCursorNC highlight groups.

The TermCursor highlight group is now used in the default `'guicursor'` value, which has a new entry for Terminal mode. However, the TermCursorNC highlight group is no longer supported: since terminal windows now use the real cursor, when the window is not focused there is no cursor displayed in the window at all, so there is nothing to highlight. Users can still use the `StatusLineTermNC` highlight group to differentiate non-focused terminal windows.

BREAKING CHANGE: The TermCursorNC highlight group is no longer supported.

Fixes #3681.
Fixes https://github.com/neovim/neovim/issues/20325.

gpanders committed:
feat(terminal)!: cursor shape and blink

When a terminal application running inside the terminal emulator sets
the cursor shape or blink status of the cursor, update the cursor in the
parent terminal to match.

This removes the "virtual cursor" that has been in use by the terminal
emulator since the beginning. The original rationale for using the
virtual cursor was to avoid having to support additional UI methods to
change the cursor color for other (non-TUI) UIs, instead relying on the
TermCursor and TermCursorNC highlight groups.

The TermCursor highlight group is now used in the default 'guicursor'
value, which has a new entry for Terminal mode. However, the
TermCursorNC highlight group is no longer supported: since terminal
windows now use the real cursor, when the window is not focused there is
no cursor displayed in the window at all, so there is nothing to
highlight. Users can still use the StatusLineTermNC highlight group to
differentiate non-focused terminal windows.

BREAKING CHANGE: The TermCursorNC highlight group is no longer supported.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -281,6 +281,8 @@ TERMINAL
 • The |terminal| now supports OSC 8 escape sequences and will display
   hyperlinks in supporting host terminals.
+• The |terminal| now updates the cursor shape and blink when set by an
+  application.
```
should we also mention something like "defers to the actual UI cursor instead of drawing a fake cursor"

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -45,6 +45,7 @@ cursorentry_T shape_table[SHAPE_IDX_COUNT] = {
   { "more", 0, 0, 0,   0,   0,   0, 0, 0, "m", SHAPE_MOUSE },
   { "more_lastline", 0, 0, 0,   0,   0,   0, 0, 0, "ml", SHAPE_MOUSE },
   { "showmatch", 0, 0, 0, 100, 100, 100, 0, 0, "sm", SHAPE_CURSOR },
+  { "terminal", 0, 0, 0, 0, 0, 0, 0, 0, "te", SHAPE_CURSOR },
```
Just for reference: interesting that vim doesn't have this https://github.com/vim/vim/blob/c942f84aadffd0c8969ecf81e3e9103722b2714f/src/misc2.c#L2042

gpanders replied:
That's not surprising, we are adding this here only so that we can communicate cursor shape information to the UI in the `mode_info_set` event. Since Vim doesn't have the UI architecture it makes sense they would have another way of implementing this.

justinmk commented on src/nvim/cursor_shape.c:
```diff
@@ -45,6 +45,7 @@ cursorentry_T shape_table[SHAPE_IDX_COUNT] = {
   { "more", 0, 0, 0,   0,   0,   0, 0, 0, "m", SHAPE_MOUSE },
   { "more_lastline", 0, 0, 0,   0,   0,   0, 0, 0, "ml", SHAPE_MOUSE },
   { "showmatch", 0, 0, 0, 100, 100, 100, 0, 0, "sm", SHAPE_CURSOR },
+  { "terminal", 0, 0, 0, 0, 0, 0, 0, 0, "te", SHAPE_CURSOR },
```
the `mode()` representation of terminal-mode is "t", does it make sense to match that here? 

```suggestion
  { "terminal", 0, 0, 0, 0, 0, 0, 0, 0, "t", SHAPE_CURSOR },
```

justinmk commented on src/nvim/terminal.c:
```diff
@@ -615,15 +620,16 @@ bool terminal_enter(void)
   curwin->w_redr_status = true;  // For mode() in statusline. #8323
   redraw_custom_title_later();
-  ui_busy_start();
+  ui_cursor_shape();
```
for GUIs should news.txt mention something like "terminal-mode no longer signals 'busy' event" ? cc @fredizzimo

gpanders replied:
AFAICT this was only being used to hide the cursor in the TUI. I'm not sure how GUIs might have been using it though.

justinmk reviewed (approved):
Very excited for this!

cosminadrianpopescu commented:
I am trying to test it and the cursor seems to be always one character to the left while editing. Having, for example:

$> ls |-

I would expect in this situation, when i click "a", the a to appear before the "-" character, but it appears after. See the attached movie.
[capture.webm](https://github.com/user-attachments/assets/416a876f-a034-4963-b4cb-807fcf4c4915)

gpanders commented:
> I am trying to test it and the cursor seems to be always one character to the left while editing. Having, for example:
> 
> $> ls |-
> 
> I would expect in this situation, when i click "a", the a to appear before the "-" character, but it appears after. See the attached movie. [capture.webm](https://github.com/user-attachments/assets/416a876f-a034-4963-b4cb-807fcf4c4915)

This is mentioned in the PR description:

>There is one bug I'm chasing down where exiting and then re-entering Terminal mode causes a disagreement in cursor position between Nvim and the terminal. But otherwise it seems to be working ok.

gpanders commented:
I was able to solve the cursor issue so this is fully working now. I also implemented cursor hiding, so if an application in the terminal emulator hides the cursor it will be hidden in Nvim too.

Because we are using the real cursor now and not the virtual cursor, I had to update any test that uses the terminal cursor in the expected screen output... which was a lot of them. There are some tests that fail now because of TermCursor. I've marked those as `pending` for now.

justinmk commented on runtime/doc/news.txt:
```diff
@@ -281,6 +281,8 @@ TERMINAL
   'scrollback' are not reflown.
 • The |terminal| now supports OSC 8 escape sequences and will display
   hyperlinks in supporting host terminals.
+• The |terminal| now updates the cursor shape and blink when set by an
```
we should also mention `TermCursor` and `TermCursorNC` highlights currently don't work. maybe in the `:help news-breaking-dev` if we expect to fix this before 0.11 release.

gpanders replied:
I fixed `TermCursor`, so we won't regress that in this PR (it was easy, just set it in `'guicursor'` and everything magically worked).

`TermCursorNC` will no longer be supported at all though. This is not a temporary regression, but a permanent change. When the terminal window doesn't have focus, there is simply no cursor, so there's no highlight group to apply. We still have `StatusLineTermNC` though.

I'll add this to `news.txt` and `deprecated.txt`.

justinmk commented on test/functional/ui/cursor_spec.lua:
```diff
@@ -190,6 +190,19 @@ describe('ui/cursor', function()
+        id_lm = 0,
+        attr = {},
+        attr_lm = {},
+        short_name = 't',
```
does `:help 'guicursor'` need an update?

gpanders replied:
Yes, done.

zeertzjq commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -248,8 +248,21 @@ it('CursorLine and CursorColumn work in :terminal buffer in Normal mode', functi
-  -- CursorLine and CursorColumn are combined with TermCursorNC.
+end)
+
+-- Pending until TermCursor highlights are working again #31562
+pending('CursorLine and CursorColumn are combined with TermCursorNC', function()
```
This is a test for the drawline.c terminal highlight combining code. If the terminal cursor is no longer highlighted by drawline.c, some other highlights should be added to the terminal.

zeertzjq commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -248,8 +248,9 @@ it('CursorLine and CursorColumn work in :terminal buffer in Normal mode', functi
   ]])
-  -- CursorLine and CursorColumn are combined with TermCursorNC.
-  command('highlight TermCursorNC gui=reverse')
+  -- CursorLine and CursorColumn are combined with other highlight groups
+  command('highlight Search NONE gui=reverse')
+  command('match Search /\\%6l\\%11c/')
```
Match highlight does not come from the terminal and does not cover the drawline.c terminal highlight code. The highlights need to come from the program running in the terminal.

gpanders replied:
See https://github.com/neovim/neovim/pull/31562#discussion_r1884787479

zeertzjq replied:
?

gpanders replied:
TermCursorNC doesn’t come from the program running in the terminal either? What is this supposed to be testing? Your PR adding this test had zero context or other information.

By the way, editing your comment after the fact and then asking if I didn’t understand it is poor form.

zeertzjq replied:
> By the way, editing your comment after the fact and then asking if I didn’t understand it is poor form.

After what fact? I only made a one-character edit to my comment before I saw your reply, and it didn't change the overall meaning.

> By “comes from the terminal” do you mean a highlight that has the `kHlTerminal` kind?

No, I mean it needs to come from `terminal_get_line_attributes()`.

gpanders replied:
>No, I mean it needs to come from terminal_get_line_attributes(). But since TermCursorNC is removed, it can now only come from the program running in the terminal.

Ok the updated test ought to be sufficient now I think. Instead of the old cursor I printed text to the terminal using different SGRs.

gpanders commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -269,8 +270,8 @@ it('CursorLine and CursorColumn work in :terminal buffer in Normal mode', functi
                                                       |
   ]])
-  -- TermCursorNC has higher precedence.
-  command('highlight TermCursorNC gui=NONE guibg=Red')
+  -- Search has higher precedence.
+  command('highlight Search gui=NONE guibg=Red')
```
@zeertzjq I updated this test to use `Search` instead of `TermCursorNC`. All of the tests pass except for the one on line 264: the highlight group on line 270 is `2` instead of `4`. I am not sure why this is the case because the test on line 254 passes correctly.

Will look at it more later, but if you have any ideas why this test is failing let me know.

gpanders commented on test/functional/terminal/cursor_spec.lua:
```diff
@@ -93,39 +93,44 @@ describe(':terminal cursor', function()
 end)
 
-describe('cursor with customized highlighting', function()
+-- Pending until screen.lua supports coloring the cursor
+pending('cursor with customized highlighting', function()
```
`TermCursor` highlights work but AFAICT screen.lua doesn't handle coloring the cursor based on hl groups from `'guicursor'`. I tried implementing this but ran out of time tonight. Will try again later.

gpanders replied:
I'm wondering if we should just remove this test. The terminal cursor color is set in `'guicursor'` now, so is colored by the UI. So this is a UI test now, not a Nvim core test, and we already have a test for `'guicolor'` in `ui/cursor_spec.lua`.

justinmk commented on runtime/doc/deprecated.txt:
```diff
@@ -66,6 +66,9 @@ LSP
 • `client.on_attach()`			Use |Client:on_attach()| instead.
 • `vim.lsp.start_client()`		Use |vim.lsp.start()| instead.
 
+HIGHLIGHTS
```
nit: sections are ordered alphabetically

justinmk commented on runtime/doc/news.txt:
```diff
@@ -219,6 +219,11 @@ EVENTS
+HIGHLIGHTS
+
+• |TermCursorNC| is removed and no longer supported. Unfocused terminals no
+  longer have a cursor.
```
is this intended to be in `news-breaking` ?

gpanders replied:
Yes, thanks

zeertzjq commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -195,7 +195,7 @@ it('CursorLine and CursorColumn work in :terminal buffer in Normal mode', functi
   screen:set_default_attr_ids({
     [1] = { background = Screen.colors.Grey90 }, -- CursorLine, CursorColumn
-    [2] = { reverse = true }, -- TermCursor
+    [2] = { reverse = true }, -- Search
```
```suggestion
    [2] = { reverse = true },
```

zeertzjq commented on test/functional/terminal/highlight_spec.lua:
```diff
@@ -248,46 +248,54 @@ it('CursorLine and CursorColumn work in :terminal buffer in Normal mode', functi
   ]])
-  -- CursorLine and CursorColumn are combined with TermCursorNC.
-  command('highlight TermCursorNC gui=reverse')
+  -- CursorLine and CursorColumn are combined with other highlight groups
```
```suggestion
  -- CursorLine and CursorColumn are combined with terminal highlights.
```

gpanders commented:
@cosminadrianpopescu Should be fixed now, can you please try again?

cosminadrianpopescu commented:
@gpanders it seems to be working now. I will test more later. But the initial tests look good. The issue dissapeared. 

Thank you for the fix and for this PR.

gpanders commented:
Windows only test failure:

```
FAILED   1 test, listed below:
FAILED   test/functional\terminal\highlight_spec.lua @ 193: CursorLine and CursorColumn work in :terminal buffer in Normal mode
test/functional\terminal\highlight_spec.lua:255: Row 6 did not match.
Expected:
  |tty ready{1: }                                        |
  | foobar f{1:o}obar foobar foobar foobar foobar foobar |
  |foobar fo{1:o}bar foobar foobar foobar foobar foobar f|
  |oobar foo{1:b}ar foobar foobar foobar foobar foobar fo|
  |obar foob{1:a}r foobar foobar foobar foobar foobar foo|
  |*{1:bar fooba^r}{4: foobar}{1:                                 }|
  |                                                  |
Actual:
  |tty ready{1: }                                        |
  | foobar f{1:o}obar foobar foobar foobar foobar foobar |
  |foobar fo{1:o}bar foobar foobar foobar foobar foobar f|
  |oobar foo{1:b}ar foobar foobar foobar foobar foobar fo|
  |obar foob{1:a}r foobar foobar foobar foobar foobar foo|
  |*{1:bar fooba^r foobar                                 }|
  |                                                  |
```

But I cannot repro locally in a Windows VM:

![image](https://github.com/user-attachments/assets/087e1a34-7920-4e21-a3fd-9b14921bf0cb)

gpanders commented:
Windows only test failure:

```
FAILED   test/functional\terminal\highlight_spec.lua @ 193: CursorLine and CursorColumn work in :terminal buffer in Normal mode
test/functional\terminal\highlight_spec.lua:255: Row 6 did not match.
Expected:
  |tty ready{1: }                                        |
  | foobar f{1:o}obar foobar foobar foobar foobar foobar |
  |foobar fo{1:o}bar foobar foobar foobar foobar foobar f|
  |oobar foo{1:b}ar foobar foobar foobar foobar foobar fo|
  |obar foob{1:a}r foobar foobar foobar foobar foobar foo|
  |*{1:bar fooba^r}{4: foobar}{1:                                 }|
  |                                                  |
Actual:
  |tty ready{1: }                                        |
  | foobar f{1:o}obar foobar foobar foobar foobar foobar |
  |foobar fo{1:o}bar foobar foobar foobar foobar foobar f|
  |oobar foo{1:b}ar foobar foobar foobar foobar foobar fo|
  |obar foob{1:a}r foobar foobar foobar foobar foobar foo|
  |*{1:bar fooba^r foobar                                 }|
  |                                                  |
```

I can reproduce this locally in a Windows VM as well. Curiously, I can't get the terminal to support any SGR sequences on Windows. Even when I run Nvim normally and try to emit an SGR (e.g. `vim.api.nvim_chan_send(vim.b.terminal_job_id, '\027[7m foobar\027[0m')`), it doesn't work.

Are SGRs known not to work on Windows for some reason? This seems unrelated to this PR so I might slap a conditional skip on this test for now.

gpanders commented:
>At this moment in the first terminal, we have a block cursor, although we should have the line cursor. The cursor will change back to line if I pass through normal mode, by doing a <C-\><C-n> and then i again. But if I don't pass through normal mode, the cursor will stay a block cursor.

Another good catch. Should be fixed now, can you please try again?

cosminadrianpopescu commented:
> Another good catch. Should be fixed now, can you please try again?

No, unfortunately it's still not working. Same behavior as previous. Did I missed something? I did a `git pull`, compile again, make install and restarted nvim,

gpanders commented:
Oh I see, your mappings change the window without exiting and Terminal mode. We need to also update the cursor if the window changes.

cosminadrianpopescu commented:
> EDIT: Should be fixed now in 9360236. Thanks for finding all these corner cases, let me know if you find any others!

Yes, it works now. Thank you very much. At the moment everything seems to be in order.

Tomorrow I will stress it some more, as I will use it a real environment, at work and I'll let you know if I find anything else. 

Again, great PR.

cosminadrianpopescu commented:
Related with the previous bug, I've noticed that if the actions are in the other order, then the cursor is lost until a new terminal is created. 

Given the mapping file (/tmp/mappings.lua):

```lua
vim.api.nvim_set_keymap('t', '<f1>', '', {
    callback = function()
        vim.api.nvim_command('tabn 1')
        vim.api.nvim_command('startinsert')
    end
})


vim.api.nvim_set_keymap('t', '<f2>', '', {
    callback = function()
        vim.api.nvim_command('tabn 2')
        vim.api.nvim_command('startinsert')
    end
})
```

then, in a new terminal: 

```
nvim --clean<cr>
:luafile /tmp/mappings.lua<cr>
:terminal<cr>i
<C-\><C-n>
:tabnew<cr>
:terminal<cr> # the line cursor is visible now
vifm<cr> # there is no cursor now
<f1> # still no cursor in the first tab
<C-\><C-n>i # still no cursor
```

It appears that there is no way to get back the cursor, after this. Even by doing `echo -e '\e[5 q'` or `reset` will not bring back the cursor. Maybe this helps.

gpanders commented:
I believe I've fixed the issues with cursor visibility (and I've added tests for those as well).

I can repro your issue with the nested `nvim` instance as well. I'm not sure exactly what's happening there so I'll have to dig into that a bit more.

cosminadrianpopescu commented:
Thank you for the latest fixes. I can confirm that indeed, the cursor visibility issue is fixed. Let me know when you want me to test nvim/vim nested issue.

gpanders commented:
@cosminadrianpopescu Please try again, I've pushed a fix that at least addresses the reproduction you gave me.

cosminadrianpopescu commented:
> @cosminadrianpopescu Please try again, I've pushed a fix that at least addresses the reproduction you gave me.

The nvim nested issue seems to be fixed. 

However, the previous issue (when the cursor is hidden by apps like `vifm`) seems I could reproduce it again. 

If I follow the steps I gave you, then everything is fine. But I manage to still encounter the issue using some scripts to send keys to the terminal. I need to isolate it and find a reproducible scenario to pass you with a `nvim --clean` and I'll be back with that issue.

gpanders commented:
@cosminadrianpopescu Since we’re now just looking at some rarer corner cases, I’m going to move forward merging this. Please open a new issue with your reproduction steps when you’re able.

cosminadrianpopescu commented:
> @cosminadrianpopescu Since we’re now just looking at some rarer corner cases, I’m going to move forward merging this. Please open a new issue with your reproduction steps when you’re able.

Sure. 

However, at the moment I'm not able to reproduce with a clean nvim. I tried all this morning and I didn't manage to do it. I still have to identify the issue on my config which is still generating the issue. But normal usage, without sending keys programmatically to the terminal and without other plugins (I also have 4 external plugins in my config) does not seem to produce the issue.

lifei commented:
Is this feature enable in v0.10.3?

lifei commented:
> > Is this feature enable in v0.10.3?
> 
> No, we do not usually backport breaking changes.

Thank you for quick replying. Which version will this feature enabled in?

zeertzjq commented:
0.11
